x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -0,0 +1,353 @@
+  op?: OperationType;
+  field?: IndexPatternField;
+}): IndexPatternLayer {
+  const previousColumn = layer.columns[columnId];
+  if (!previousColumn) {
+    throw new Error(`Can't replace column because there is no prior column`);
+  }
+
+  const isNewOperation = Boolean(op) && op !== previousColumn.operationType;
+  const operationDefinition = operationDefinitionMap[op || previousColumn.operationType];
+
+  if (!operationDefinition) {
+    throw new Error('No suitable operation found for given parameters');
+  }
+
+  const baseOptions = {
+    columns: layer.columns,
+    indexPattern,
+    previousColumn: layer.columns[columnId],
+  };
+
+  if (isNewOperation) {
+    // TODO: Reference based operations require more setup to create the references
+
+    if (operationDefinition.input === 'none') {
+      const newColumn = operationDefinition.buildColumn(baseOptions);
+
+      if (previousColumn.customLabel) {
+        newColumn.customLabel = true;
+        newColumn.label = previousColumn.label;
+      }
+
+      return {
+        ...layer,
+        columns: adjustColumnReferencesForChangedColumn(
+          { ...layer.columns, [columnId]: newColumn },
+          columnId
+        ),
+      };
+    }
+
+    if (!field) {
+      throw new Error(`Invariant error: ${operationDefinition.type} operation requires field`);
+    }
+
+    const newColumn = operationDefinition.buildColumn({ ...baseOptions, field });
+
+    if (previousColumn.customLabel) {
+      newColumn.customLabel = true;
+      newColumn.label = previousColumn.label;
+    }
+
+    const newColumns = { ...layer.columns, [columnId]: newColumn };
+    return {
+      ...layer,
+      columnOrder: getColumnOrder({ ...layer, columns: newColumns }),
+      columns: adjustColumnReferencesForChangedColumn(newColumns, columnId),
+    };
+  } else if (
+    operationDefinition.input === 'field' &&
+    field &&
+    ('sourceField' in previousColumn ? field.name !== previousColumn.sourceField : true)
+  ) {
+    const newColumn = operationDefinition.buildColumn({ ...baseOptions, field, previousColumn });
+
+    // TODO: This is probably wrong, not all params are shared on all ops?
+    if ('params' in previousColumn) {
+      newColumn.params = previousColumn.params;
+    }
+
+    if (previousColumn.customLabel) {
+      newColumn.customLabel = true;
+      newColumn.label = previousColumn.label;
+    }
+
+    const newColumns = { ...layer.columns, [columnId]: newColumn };
+    return {
+      ...layer,
+      columnOrder: getColumnOrder({ ...layer, columns: newColumns }),
+      columns: adjustColumnReferencesForChangedColumn(newColumns, columnId),
+    };
+  } else {
+    throw new Error('nothing changed');
+  }
+}
+
+function addBucket(
+  layer: IndexPatternLayer,
+  column: IndexPatternColumn,
+  addedColumnId: string
+): IndexPatternLayer {
+  const [buckets, metrics] = separateBucketColumns(layer);
+
+  if (buckets.length === 0 && column.operationType === 'terms') {
```

Comment:
No need to keep this logic around in the generic helper - `buildColumns` already gets a map of all current columns, so it can implement this logic itself.

Reply:
Good point, done